File: code/nest/lab3/l3e4_global_page_management_kernel.cc
#include "machine.h"
#include "vm.h"

bool occupied[NumPhysPages];

int Nest(void *arg) 
{
    for (int i = 1; i <= NumPhysPages+1; i++)
    {
        int page = get_and_use_page();
        DEBUG('e', "get_and_use_page() returns %d\n", page);
        if (page < 0)
        {
            if (i == NumPhysPages + 1)
            {
                continue;
            }
            else
            {
                for (int j = 0; j < NumPhysPages; j++)
                {
                    if (!occupied[j])
                    {
                        DEBUG('e', "get_and_use_page() returned %d, but page %d is still unoccupied.\n", page, j);
                        // return false;
                        break;
                    }
                }
            }
        }
        else
        {
            if (page >= NumPhysPages)
            {
                DEBUG('e', "Page returned by get_and_use_page() should be less than NumPhysPages(%d).\n", NumPhysPages);
                // return false;
            }
            if (occupied[page])
            {
                DEBUG('e', "get_and_use_page() returned %d, but it has been occupied.\n", page);
                // return false
            }
            occupied[page] = true;
        }
    }

    int free_page_num = 0;
    for (int i = 1; i <= NumPhysPages; i += 3)
    {
        free_page(i);
        occupied[i] = false;
        free_page_num++;
        DEBUG('e', "free_page(%d).\n", i);
    }

    for (int i = 1; i <= free_page_num+1; i++)
    {
        int page = get_and_use_page();
        DEBUG('e', "get_and_use_page() returns %d\n", page);
        if (page < 0)
        {
            if (i == NumPhysPages + 1)
            {
                continue;
            }
            else
            {
                for (int j = 0; j < NumPhysPages; j++)
                {
                    if (!occupied[j])
                    {
                        DEBUG('e', "get_and_use_page() returned %d, but page %d is still unoccupied.\n", page, j);
                        // return false;
                        break;
                    }
                }
            }
        }
        else
        {
            if (page >= NumPhysPages)
            {
                DEBUG('e', "Page returned by get_and_use_page() should be less than NumPhysPages(%d).\n", NumPhysPages);
                // return false;
            }
            if (occupied[page])
            {
                DEBUG('e', "get_and_use_page() returned %d, but it has been occupied.\n", page);
                // return false;
            }
            occupied[page] = true;
        }
    }
    return 0;
}

File: code/nest/lab3/l3e4_global_page_management_kernel.py
import re
from collections import OrderedDict
from parse_result import *
from operator import attrgetter

def is_false_failure(line):
    a = re.findall('get_and_use_page\(\) returned (.*), but page (.*) is still unoccupied.\n', line)
    if (len(a) > 0):
        return int(a[0][1])
    else:
        return None

def is_page_too_large(line):
    a = re.findall('Page returned by get_and_use_page\(\) should be less than NumPhysPages(.*).\n', line)
    if (len(a) > 0):
        return True
    else:
        return None

def is_page_occupied(line):
    a = re.findall('get_and_use_page\(\) returned (.*), but it has been occupied.\n', line)
    if (len(a) > 0):
        return True
    else:
        return None

def check(lines):
    exit_successfully = False
    for line_num, line in enumerate(lines):
        ret = is_false_failure(line)
        if ret is not None:
            return False, 'Getting page failed, but there are pages unoccupied'
            
        ret = is_page_too_large(line)
        if ret is not None:
            return False, 'Page returned is not right'
            
        ret = is_page_occupied(line)
        if ret is not None:
            return False, 'Page returned has been occupied'
    
    return True, None


